### ISSUE TITLE
Environment variable specifying search paths for scripts

### ISSUE THREAD

**Issue by arunisaac** [weechat/weechat#1716]
## Feature description

It would be nice to have an environment variable, say `WEECHAT_SCRIPTS_PATH`, that specifies a colon separated list of paths in which to search for scripts in addition to `~/.local/share/weechat/...` and `/usr/share/weechat/...`. I need this because I am trying to package several weechat scripts for GNU Guix. Guix does not install all packages to standard locations like `/usr/share/weechat/...`. Instead it installs each package into a separate directory under `/gnu/store`. Having such an environment variable would obviate the need for the user to manually create a symlink to the script, and would thus improve the user experience.

Thank you!

**Comment by ksu-0** [comment]
Why can't you make a `usr/share/weechat` link farm under Guix profile, like other packages do?

**Comment by arunisaac** [comment]
Hi,

> Why can't you make a `usr/share/weechat` link farm under Guix profile,

A link farm wouldn't work all the time. Guix profiles can exist on
multiple levels. When packages are installed to the global system
profile, a link farm is created under
/run/current-system/profile/share/weechat. When packages are installed
to the user profile, a link farm is created under
~/.guix-profile/share/weechat. When packages are installed to a
temporary profile (as created by `guix environment`), a link farm is
created under /gnu/store/...-profile/share/weechat.

Hardcoding weechat to search under
/run/current-system/profile/share/weechat and
~/.guix-profile/share/weechat would not cover the third use case of
/gnu/store/...-profile/share/weechat. Here the ellipsis (...) is a hash
that is not known at weechat compile time.

> like other packages do?

Do you know any packages that manage to do this without an environment
variable? If you could point them to me, I could take a look and see how
they do it.

Thanks!

**Comment by ksu-0** [comment]
Thank you, it wasn't obvious to me from the original request that the link farm is only part of the solution.

**Comment by flashcode** [comment]
@arunisaac: yes such variable can be added, should this path have a higher priority when searching where is the script to load? (commonly env vars are used to force something over a standard behavior, so I would say, it's used first if the env var is set).

**Comment by arunisaac** [comment]
Yes, you're right, it should have higher priority.

**Comment by flashcode** [comment]
I'm wondering if these paths should be used to autoload scripts as well?
Or if like sharedir, we assume autoload are only from autoload directories in the home, pointing to any path, even outside home?
Because WeeChat can not create symbolic links in system directories.

If yes, for example if `$WEECHAT_SCRIPTS_DIR` == `/tmp1:/tmp2`, such paths would be used in this order to autoload scripts on startup (by each scripting language: python, perl, etc.):

    /tmp1/{language}/autoload
    /tmp2/{language}/autoload
    /home/user/.local/share/weechat/{language}/autoload

And when loading manually a script (eg: `/script load foo.py` or `/python load foo.py`), these paths would be used in this order:

    /tmp1/{language}/autoload
    /tmp1/{language}
    /tmp2/{language}/autoload
    /tmp2/{language}
    /home/user/.local/share/weechat/{language}/autoload
    /home/user/.local/share/weechat/{language}
    /home/user/.local/share/weechat
    /usr/share/weechat/{language}

If autoload is only in WeeChat data dir, the paths would be used in this order when loading manually a script:

    /tmp1/{language}
    /tmp2/{language}
    /home/user/.local/share/weechat/{language}/autoload
    /home/user/.local/share/weechat/{language}
    /home/user/.local/share/weechat
    /usr/share/weechat/{language}

**Comment by arunisaac** [comment]
> I'm wondering if these paths should be used to autoload scripts as well?
> Or if like sharedir, we assume autoload are only from autoload directories in the home, pointing to any path, even outside home?
> Because WeeChat can not create symbolic links in system directories.

I think the proposed WEECHAT_SCRIPTS_DIR environment variable should be
used to autoload scripts as well. WeeChat may not be able to create
symlinks in system directories, but Guix (and other distributions)
might.

If need be, we can also isolate the autoload and manual load
functionality into two separate environment variables---say,
WEECHAT_SCRIPTS_DIR and WEECHAT_AUTOLOAD_SCRIPTS_DIR.

**Comment by flashcode** [comment]
Then if it is used for autoloading as well, what should be done there:

- creation of autoload directorires on startup: this is done only in home for now
- when a script is installed, the script is downloaded in home, and a symlink is made from autoload (in home) to the script (in home as well).

Having another location for scripts and autoload of scripts could be confusing, the script plugin should always download and create symlinks in a single location for me.

So I'm wondering if such env var WEECHAT_SCRIPTS_PATH should override the default scripts location, if set (ie use only this and never the scripts in WeeChat data directory).

These points must be clarified, I started some developments, but I'll wait for complete specs before going on. For now I remove the target milestone.

**Comment by arunisaac** [comment]
The idea is that we support two use cases:

- one where weechat manages download and installation of scripts into
home. We already have this.

- the second where an external package manager (such as Guix) manages
scripts in system directories. In this case, each script would be a
separate Guix package. This use case is important because it allows the
user to use one centralized package manager for all their packages
(regular packages, weechat plugins, python packages, emacs packages or
whatever)

The script plugin should always download and create symlinks in home. In
any case, it won't have permissions to download to system
directories. Management of scripts, creation of autoload directories,
etc. in system directories should be handled only by the package manager
(such as Guix). The WEECHAT_SCRIPTS_PATH environment variable should
however search in both home and in system directories (with higher
priority to home). This supports the use case in which the user decides
to have some scripts installed system-wide and some only in home.

Hope that was helpful. Do let me know if you have any questions.

> These points must be clarified, I started some developments, but I'll
> wait for complete specs before going on. For now I remove the target
> milestone.

Thanks for working on this!

**Comment by NMerkley** [comment]
Is there any reason to have a weechat specific environment variable for this instead of just searching XDG_DATA_DIRS? The current behavior of looking at .local/share and /usr/share is basically doing that already just hardcoded

**Comment by arunisaac** [comment]
Ah, you're right.

Guix does not currently set up XDG_DATA_DIRS for weechat. But, I could
modify the weechat package definition so that it does. In that case, a
WEECHAT_SCRIPTS_PATH would indeed not be necessary.

**Comment by NMerkley** [comment]
I believe that Guix does the proper thing here but Weechat is not searching XDG_DATA_DIRS. I stumbled on this issue yesterday trying to install a Weechat script on Guix. It appeared in my profile properly, but I had to symlink the guix profile share/weechat/python folder to .local/share/weechat/python

**Comment by flashcode** [comment]
Yes I confirm this is not yet supported, as written in the spec on XDG support: https://specs.weechat.org/specs/2021-001-follow-xdg-base-dir-spec.html#out-of-scope
